static/js/plugin_loading_and_search.js

```diff
@@ -0,0 +1,169 @@
+      .getElementById("pluginContainer")
+      .getElementsByTagName("a");
+
+    // Filter the list of all <a> tags to only those that are currently displayed
+    const visiblePlugins = Array.from(allPlugins).filter(
+      (plugin) => plugin.style.display !== "none",
+    );
+
+    // Update the displayed count with the number of visible plugins
+    document.getElementById("pluginCount").textContent = visiblePlugins.length;
+  }
+
+  let searchTimeout;
+
+  function attachSearchListener() {
+    const searchBox = document.getElementById("searchBox");
+    searchBox.addEventListener("input", function (event) {
+      clearTimeout(searchTimeout);
+      searchTimeout = setTimeout(() => {
+        handleSearch(event.target.value);
+      }, 300);
+    });
+
+    // trigger search if user already typed something
+    if (searchBox.value.trim()) {
+      handleSearch(searchBox.value);
+    }
+  }
+
+  function makeSpinner(text) {
+    const container = document.createElement("div");
+    container.className = "spinner-container";
+    const spinner = document.createElement("span");
+    spinner.className = "spinner";
+    const textNode = document.createElement("span");
+    textNode.textContent = text;
+    container.appendChild(spinner);
+    container.appendChild(textNode);
+    return container;
+  }
+
+  async function handleSearch(searchText) {
+    // cancel previous search
+    if (currentSearchAbortController) {
+      currentSearchAbortController.abort();
+    }
+
+    // if search is empty, show all plugins
+    if (!searchText.trim()) {
+      const pluginContainer = document.getElementById("pluginContainer");
+      const existingSpinner =
+        pluginContainer.querySelector(".spinner-container");
+      if (existingSpinner) {
+        pluginContainer.removeChild(existingSpinner);
+      }
+      for (const plugin of pluginContainer.children) {
+        plugin.style.display = "block";
+      }
+      updatePluginCount();
+      return;
+    }
+
+    // create new controller for this search
+    currentSearchAbortController = new AbortController();
+
+    try {
+      await streamSearchResults(
+        searchText,
+        currentSearchAbortController.signal,
+      );
+    } catch (error) {
+      if (error.name !== "AbortError") {
+        console.error("Search error:", error);
+      }
+    }
+  }
+
+  async function streamSearchResults(searchText, signal) {
+    // hide all plugins initially
+    const pluginContainer = document.getElementById("pluginContainer");
+    for (const plugin of pluginContainer.children) {
+      plugin.style.display = "none";
+    }
+
+    // remove any existing spinner first
+    const existingSpinner = pluginContainer.querySelector(".spinner-container");
+    if (existingSpinner) {
+      pluginContainer.removeChild(existingSpinner);
+    }
+
+    const spinner = makeSpinner("Searching plugins...");
+    pluginContainer.insertBefore(spinner, pluginContainer.firstChild);
```

Comment:
Is this designed to push the spinner to the bottom, underneath the search results? I think that would be nice, but I'm seeing it above the search results. I think this is not a deal-breaker regardless, but just wanted to double-check for my own understanding.

Reply:
I wanted the spinner to stay visible while results come in, to indicate they are not complete. If it's at the bottom it can quickly be pushed off the screen.

I moved the spinner to be inside the search bar to avoid weird layout shifts, but we can revisit this design!